Dockerfile

```diff
@@ -1,71 +1,89 @@
 
-# Install bundler to run bundle exec
-# This should be the same version as the Gemfile.lock
+# Bundler version should be the same version as what the Gemfile.lock was bundled with
 RUN gem install bundler:2.3.14 --no-document
 
-# Install gems defined in Gemfile
 COPY .ruby-version Gemfile Gemfile.lock /app/
-
 RUN bundle config set without "development test"
-ARG BUNDLE_FLAGS="--jobs=4 --no-binstubs --no-cache"
-RUN bundle install ${BUNDLE_FLAGS}
+RUN bundle install --jobs=4 --no-binstubs --no-cache
 
-# Install node packages defined in package.json, including webpack
 COPY package.json yarn.lock /app/
 RUN yarn install --frozen-lockfile
 
-# Copy all files to /app (except what is defined in .dockerignore)
 COPY . /app/
 
-# Compile assets and run webpack. We set a dummy secret key.
-RUN RAILS_ENV=production bundle exec rails SECRET_KEY_BASE=pickasecuretoken assets:precompile
+RUN bundle exec rake assets:precompile
 
-# Cleanup to save space in the production image
-RUN rm -rf node_modules log tmp && \
-      rm -rf /usr/local/bundle/cache && \
-      rm -rf .env && \
-      find /usr/local/bundle/gems -name "*.c" -delete && \
-      find /usr/local/bundle/gems -name "*.h" -delete && \
-      find /usr/local/bundle/gems -name "*.o" -delete && \
-      find /usr/local/bundle/gems -name "*.html" -delete
+ENV PORT=8080
+EXPOSE ${PORT}
 
-# Build runtime image
-FROM ruby:3.1.4-alpine as production
+RUN adduser --system --no-create-home nonroot
 
-# The application runs from /app
-WORKDIR /app
+# We expect the rake assets:precompile command to create these directories, but mkdir -p will create them if they don't already exist
+RUN mkdir -p tmp log
+RUN chown -R nonroot tmp log
+RUN chown nonroot db/schema.rb
 
-# Add postgres driver library
-# Add the timezone as it's not configured by default in Alpine
-RUN apk add --update --no-cache libpq tzdata && \
-  cp /usr/share/zoneinfo/Europe/London /etc/localtime && \
-  echo "Europe/London" > /etc/timezone
+FROM base as development
 
-# Copy files generated in the builder image
-COPY --from=builder /app /app
-COPY --from=builder /usr/local/bundle/ /usr/local/bundle/
+RUN bundle config set without ""
+RUN bundle install --jobs=4 --no-binstubs --no-cache
 
-ARG GIT_COMMIT_SHA="UNKNOWN"
-ENV GIT_COMMIT_SHA=${GIT_COMMIT_SHA}
-RUN echo ${GIT_COMMIT_SHA} > ./GIT_COMMIT_SHA
+# Install gecko driver for Capybara tests
+RUN apk add firefox
+RUN wget https://github.com/mozilla/geckodriver/releases/download/v0.31.0/geckodriver-v0.31.0-linux64.tar.gz \
+    && tar -xvzf geckodriver-v0.31.0-linux64.tar.gz \
+    && rm geckodriver-v0.31.0-linux64.tar.gz \
+    && chmod +x geckodriver \
+    && mv geckodriver /usr/local/bin/
 
-ENV PORT=8080
+USER nonroot
 
-EXPOSE ${PORT}
+CMD bundle exec rails s -e ${RAILS_ENV} -p ${PORT} --binding=0.0.0.0
+
+FROM base as staging
+
+RUN bundle config set without "development"
+RUN bundle install --jobs=4 --no-binstubs --no-cache
```

Comment:
Why is this necessary for "staging" and not production?

It would be nice if you could have the same image for both staging and prod.

Reply:
If you can, then using the same image on both staging and production (the only difference being env vars) would reduce a possible source of "but it worked on staging!" surprises. In this particular case, is there any chance of bundler having issues with nested dependencies conflicting for example?

I don't see why you need e.g. selenium drivers installed on staging so I'd just remove it, simplify this file a bit and you have to publish fewer docker images

Reply:
We've updated this now to remove the `staging` stage. The staging and prod apps will both be built from the `prod` stage of the Dockerfile.